include/swift/Syntax/AbsoluteRawSyntax.h

```diff
@@ -218,37 +218,144 @@ struct AbsoluteRawSyntax {
   const AbsoluteSyntaxInfo Info;
 
 public:
+  /// Create a null \c AbsoluteRawSyntax to which a real \c AbsoluteRawSyntax
+  /// can be stored later.
+  explicit AbsoluteRawSyntax()
+      : Raw(nullptr), Info(AbsoluteSyntaxPosition(0, 0),
+                           SyntaxIdentifier(0, SyntaxIndexInTree::zero())) {}
+
+  /// Create a new \c AbsoluteRawData backed by \p Raw and with additional \p
+  /// Info. The caller of this constructor is responsible to ensure that the
+  /// Arena of \p Raw (and thus \p Raw itself) outlives this \c
+  /// AbsoluteRawSyntax.
   AbsoluteRawSyntax(const RawSyntax *Raw, AbsoluteSyntaxInfo Info)
-      : Raw(Raw), Info(Info) {}
+      : Raw(Raw), Info(Info) {
+    assert(Raw != nullptr &&
+           "A AbsoluteRawSyntax created through the memberwise constructor "
+           "should always have a RawSyntax");
+  }
+
+  /// Whether this is a null \c AbsoluteRawSyntax created through the default
+  /// constructor.
+  bool isNull() const { return Raw == nullptr; }
 
   /// Construct a \c AbsoluteRawSyntax for a \c RawSyntax node that represents
   /// the syntax tree's root.
   static AbsoluteRawSyntax forRoot(const RawSyntax *Raw) {
     return AbsoluteRawSyntax(Raw, AbsoluteSyntaxInfo::forRoot());
   }
 
-  const RawSyntax *getRaw() const { return Raw; }
+  const RawSyntax *getRaw() const {
+    assert(!isNull() && "Cannot get Raw of a null AbsoluteRawSyntax");
+    return Raw;
+  }
 
-  AbsoluteSyntaxInfo getInfo() const { return Info; }
+  AbsoluteSyntaxInfo getInfo() const {
+    assert(!isNull() && "Cannot get Raw of a null AbsoluteRawSyntax");
+    return Info;
+  }
 
   /// Get the position at which the leading triva of this node starts.
-  AbsoluteSyntaxPosition getPosition() const { return Info.getPosition(); };
+  AbsoluteSyntaxPosition getPosition() const {
+    return getInfo().getPosition();
+  };
 
-  SyntaxIdentifier getNodeId() const { return Info.getNodeId(); };
+  SyntaxIdentifier getNodeId() const { return getInfo().getNodeId(); };
 
   AbsoluteSyntaxPosition::IndexInParentType getIndexInParent() const {
     return getPosition().getIndexInParent();
   }
 
+  size_t getNumChildren() const { return getRaw()->getLayout().size(); }
+
+  /// Get the child at \p Index if it exists. If the node does not have a child
+  /// at \p Index, return \c None. Asserts that \p Index < \c NumChildren
+  Optional<AbsoluteRawSyntax>
+  getChild(AbsoluteSyntaxPosition::IndexInParentType Index) const {
```

Comment:
Can we return `isNull()` AbsoluteRawSyntax? If not, what the purpose of `isNull()`?

Reply:
The purpose of the `isNull()` `AbsoluteRawSyntax` is so we can construct a null `SyntaxDataRef` (which has a `AbsoluteRawSyntax` member). We need this for the `getChildRef` method, I already described [here](https://github.com/apple/swift/pull/36250#issuecomment-790452915), which is used as follows:

```cpp
SyntaxRef result;
someNode.getChildRef(1, result);
```

Technically, `result` just needs to be uninitialized memory. If you know how to create uninitialized stack memory, we can remove `isNull` altogether.

I though about returning a null `AbsoluteRawSyntax` here, but decided against it, because I expect there will be a whole category of bugs where we would be accessing an optional child without check if it is null.

Reply:
I’m looking into whether we can make `Optional<AbsoluteRawSyntax>` a zero-cost wrapper around `AbsoluteRawSyntax` whose null value is internally backed by `isNull()` and make `isNull()` private. Then we would have the compile time safety of `Optional` but the runtime behavior as if we returned an `isNull()` `AbsoluteRawSyntax`.

Reply:
I removed the `isNull` function and made `Optional<AbsoluteRawSyntax>` a zero-cost wrapper. Since the change wasn't trivial (~250 LOC), I created a follow-up PR for it [here](https://github.com/apple/swift/pull/36313).